pkg/resto/resto.go

```diff
@@ -0,0 +1,357 @@
+package resto
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/cli/cli/config"
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest"
+	"github.com/docker/distribution/manifest/schema2"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/distribution/registry/client"
+	digest "github.com/opencontainers/go-digest"
+	ociv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type unsupportedMediaType struct{}
+
+func (u unsupportedMediaType) Error() string {
+	return "Unsupported media type"
+}
+
+// ManifestAny is a manifest type for arbitrary configuration data
+type ManifestAny struct {
+	manifest.Versioned
+	Payload string `json:"payload,omitempty"`
+}
+
+type parsedReference struct {
+	domain string
+	path   string
+	tag    string
+}
+
+func parseRef(repoTag string) (parsedReference, error) {
+	rawref, err := reference.ParseNormalizedNamed(repoTag)
+	if err != nil {
+		return parsedReference{}, err
+	}
+	ref, ok := rawref.(reference.Named)
+	if !ok {
+		return parseRef("docker.io/" + repoTag)
+	}
+	tag := "latest"
+	if rt, ok := ref.(reference.Tagged); ok {
+		tag = rt.Tag()
+	}
+	domain := reference.Domain(ref)
+	if domain == "docker.io" {
+		domain = "registry-1.docker.io"
+	}
+	return parsedReference{"https://" + domain, reference.Path(ref), tag}, nil
+}
+
+func getCredentials(domain string) (string, string, error) { //nolint:unparam
+	configPath := filepath.Join(config.Dir(), "config.json")
+	configFile := configfile.New(configPath)
+	cr, err := os.Open(configPath)
+	if err != nil {
+		return "", "", err
+	}
+	err = configFile.LoadFromReader(cr)
+	if err != nil {
+		return "", "", err
+	}
+	cr.Close()
+	if domain == "https://registry-1.docker.io" {
+		domain = "https://index.docker.io/v1/"
+	}
+	domain = strings.TrimPrefix(domain, "https://")
+	auth, err := configFile.GetAuthConfig(domain)
+	if err != nil {
+		//fmt.Printf("GetAuthConfig failure for %s: %s\n", domain, err)
+		return "", "", err
+	}
+	return auth.Username, auth.Password, nil
+}
+
+func makeTarGz(content map[string]string) ([]byte, digest.Digest, error) {
```

Comment:
Couldn't we use what `docker/docker` provide and uses (i.e. `pkg/archive`) and what's done in `docker/docker/distribution` for pushing code.

Reply:
Won't make things any simpler from what I see of docker/docker/distribution's interface